File: web/src/gf_apps/gf_crawl_lib/ts/gf_crawl_stats/gf_crawl_stats__links.ts
/*
GloFlow application and media management/publishing platform
Copyright (C) 2019 Ivan Trajkovic

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

///<reference path="../../../../d/jquery.d.ts" />

declare var sigma;
declare var c3;
//---------------------------------------------------
export function view__new_links_per_day(p_stats_lst, p_parent, p_log_fun) {
    p_log_fun('FUN_ENTER','gf_crawl_stats__links.view__new_links_per_day()');

    const plot = $(`
        <div id='plots'>
            <div id='new_links_per_day__plot'>
                <svg width='800' height='600'></svg>
            </div>
            <div id='total_links_count_per_day__plot'>
                <svg width='800' height='600'></svg>
            </div>
        </div>`);

    $(p_parent).append(plot);
    //-----------------
    const daily_total_count_lst           = [];
    const daily_valid_for_crawl_total_lst = [];
    const daily_fetched_total_lst         = [];
    const total_counts_lst                = [];
    
    var i               = 0;
    var total_count_int = 0;

    for (var day_stat_map of p_stats_lst) {

        console.log('>>>>> =====================================')
        console.log(day_stat_map);

        const count_int                 = day_stat_map['total_count_int'];
        const valid_for_crawl_total_int = day_stat_map['valid_for_crawl_total_int'];
        const fetched_total_int         = day_stat_map['fetched_total_int'];

        console.log(count_int);
        console.log(valid_for_crawl_total_int)
        console.log(fetched_total_int)

        daily_total_count_lst.push(count_int);
        daily_valid_for_crawl_total_lst.push(valid_for_crawl_total_int);
        daily_fetched_total_lst.push(fetched_total_int);

        total_counts_lst.push(total_count_int); //total_count_int);

        i               += 1;
        total_count_int += count_int;
    }
    //-----------------

    daily_total_count_lst.unshift('per-day new links count'); //add column title as first element
    daily_valid_for_crawl_total_lst.unshift('per-day valid-for-crawl links count');
    daily_fetched_total_lst.unshift('per-day fetched links count');

    const chart = c3.generate({
        bindto: '#new_links_per_day__plot',
        data: {
          columns: [
            daily_total_count_lst,
            daily_valid_for_crawl_total_lst,
            daily_fetched_total_lst
            //['data1', 30, 200, 100, 400, 150, 250],
            //['data2', 50, 20, 10, 40, 15, 25]
          ]
        }
    });

    total_counts_lst.unshift('per-day total links count')
    const chart2 = c3.generate({
        bindto: '#total_links_count_per_day__plot',
        data: {
          columns: [
            total_counts_lst
          ]
        }
    });
}
//---------------------------------------------------
export function view__unresolved(p_stats_lst, p_log_fun) {
	p_log_fun('FUN_ENTER','gf_crawl_stats__links.view__unresolved()');

    const container = $(`
        <div id="unresolved_links">
            <div class="title">unresolved_links</div>
        </div>`);

    for (var d_map of p_stats_lst) {

        const origin_domain_str = d_map['origin_domain_str'];

        const domain = $(`
            <div class="origin_domain">
                <div class="title">`+origin_domain_str+`</div>
                <div class="plot_urls_references_graph_btn">graph</div>
                <div class="origin_urls"></div>
            </div>`);
        $(container).append(domain);

        const origin_urls_lst               = d_map['origin_urls_lst'];
        const a_hrefs__from_origin_urls_lst = d_map['a_hrefs__from_origin_urls_lst'];

        $(domain).find('.plot_urls_references_graph_btn').on('click',(p_e)=>{

            const graph = view__links_graph(origin_domain_str,
                origin_urls_lst,
                a_hrefs__from_origin_urls_lst,
                domain,
                p_log_fun);
        });

        for (var i=0;i<origin_urls_lst.length;i++) {

            const origin_url_str              = origin_urls_lst[i];
            const a_hrefs_from_origin_url_lst = a_hrefs__from_origin_urls_lst[i];

            $(domain).append($(`
                <div>
                    <div class="origin_url">`+origin_url_str+`</div>
                    <div class="a_hrefs_count">`+a_hrefs_from_origin_url_lst.length+`</div>
                </div>`));
        }
    }

    return container;
}
//---------------------------------------------------
export function view__links_graph(p_origin_domain_str :string,
    p_origin_urls_lst               :string[],
    p_a_hrefs__from_origin_urls_lst :string[],
    p_parent,
    p_log_fun) {
    p_log_fun('FUN_ENTER','gf_crawl_stats__links.view__links_graph()');

    const c = $(`
        <div id="graph_container">
            <div id="close_btn">x</div>
        </div>`);

    $(p_parent).append(c);
        
    const edges_lst = [];
    const nodes_lst = [];
    const nodes_map = {}; //used as a SET, to eliminate duplicates (sigmajs doesnt like duplicate nodes in the nodes_lst)

    //----------------------
    nodes_lst.push({
        id:    p_origin_domain_str,
        label: p_origin_domain_str,
        x:     0,   //Math.random(),
        y:     0,   //Math.random(),
        size:  0.5, //Math.random(),
        color: '#111'
    });
    nodes_map[p_origin_domain_str] = true;
    //----------------------

    var i=0;
    for (var origin_url_str of p_origin_urls_lst) {

        if (!(origin_url_str in nodes_map)) {
            nodes_lst.push({
                id:    origin_url_str,
                label: origin_url_str,
                x:     Math.random(),
                y:     Math.random(),
                size:  0.5, //Math.random(),
                color: '#111'
            });

            nodes_map[origin_url_str] = true;
        }
    
        const a_hrefs_from_origin_url_lst = p_a_hrefs__from_origin_urls_lst[i];

        edges_lst.push({
                id:     origin_url_str+' '+i,
                source: p_origin_domain_str,
                target: origin_url_str,
                size:   Math.random(),
                color:  '#ccc'
            }); 

        for (var j=0; j < a_hrefs_from_origin_url_lst.length; j++) {

            const a_href_from_origin_url_str = a_hrefs_from_origin_url_lst[j];
                
            if (!(a_href_from_origin_url_str in nodes_map)) {
                nodes_lst.push({
                    id:    a_href_from_origin_url_str,
                    label: 'n', //a_href_from_origin_url_str,
                    x:     Math.random(),
                    y:     Math.random(),
                    size:  0.1, //Math.random(),
                    color: '#666'
                });

                nodes_map[a_href_from_origin_url_str] = true;
            }   

            edges_lst.push({
                id:     origin_url_str + '_'+a_href_from_origin_url_str+'_'+j,
                source: origin_url_str,
                target: a_href_from_origin_url_str,
                size:   0.5, //Math.random(),
                color:  '#ccc'
            });       
        }

        i+=1;
    }

    const s = new sigma({
        graph: {
            nodes: nodes_lst,
            edges: edges_lst
        },
        container: 'graph_container'
    });

    s.startForceAtlas2({worker: true, barnesHutOptimize: false});

    //-----------------
    //CLOSE_BTN
    $(c).find('#close_btn').on('click', (p_e)=>{
        s.stopForceAtlas2({worker: true, barnesHutOptimize: false});
        $(c).remove();

    });
    //-----------------
}
//---------------------------------------------------
export function view__crawled_domains(p_domains_lst, p_log_fun) {
    p_log_fun('FUN_ENTER', 'gf_crawl_stats__links.view__crawled_domains()');

    const links_domains_e = $(`
        <div id="links_domains">
            <div class="title">crawled_links_domains</div>
        </div>`);
   
    for (var domain_map of p_domains_lst) {

        const domain_str              = domain_map['domain_str'];
        const links_count_int         = domain_map['links_count_int'];
        const creation_unix_times_lst = domain_map['creation_unix_times_lst'];
        const a_href_lst              = domain_map['a_href_lst'];
        const origin_urls_lst         = domain_map['origin_urls_lst'];
        const valid_for_crawl_lst     = domain_map['valid_for_crawl_lst'];
        const fetched_lst             = domain_map['fetched_lst'];
        const images_processed_lst    = domain_map['images_processed_lst'];

        const domain_e = $(`<div class="links_domain">
                <div class='domain_name'><a href="http://`+domain_str+`" target="_blank">`+domain_str+`</a></div>
                <div class='links_count'><span class="links_count_label">links_count - </span><span class="links_count">`+links_count_int+`</span></div>
            </div>`);
        $(links_domains_e).append(domain_e);

        for (var i=0;i<a_href_lst.length;i++) {

            const a_href_str            = a_href_lst[i];
            const valid_for_crawl_bool  = valid_for_crawl_lst[i];
            const fetched_bool          = fetched_lst[i];
            const images_processed_bool = images_processed_lst[i];

            const u = `
                <div class="a_href">
                    <a target="_blank" href="`+a_href_str+`">`+a_href_str+`</a>
                    <div class="status">
                        <span class="status_item valid_for_crawl">
                            <span style="font-weight:bold;">valid_for_crawl:</span>
                            <span class="value">`+valid_for_crawl_bool+`</span>
                        </span>
                        <span class="status_item fetched">
                            <span style="font-weight:bold;">fetched:</span>
                            <span class="value">`+fetched_bool+`</span>
                        </span>
                        <span class="status_item images_processed">
                            <span style="font-weight:bold;">images_processed:</span>
                            <span class="value">`+images_processed_bool+`</span>
                        </span>
                    </div>
                </div>`;
            $(domain_e).append(u);
        }
    }

    return links_domains_e;
}

File: go/gf_apps/gf_crawl_lib/gf_crawl_stats/gf_crawl_stats__links.go
/*
GloFlow application and media management/publishing platform
Copyright (C) 2019 Ivan Trajkovic

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

package gf_crawl_stats

import (
	"fmt"
	"time"
	"strconv"
	"sort"
	"context"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/bson"
	"github.com/gloflow/gloflow/go/gf_core"
)

//-------------------------------------------------
type Gf_stat__crawled_links_domain struct {
	Domain_str              string    `bson:"_id"                     json:"domain_str"`
	Links_count_int         int       `bson:"links_count_int"         json:"links_count_int"`
	Creation_unix_times_lst []float64 `bson:"creation_unix_times_lst" json:"creation_unix_times_lst"`
	A_href_lst              []string  `bson:"a_href_lst"              json:"a_href_lst"`
	Origin_urls_lst         []string  `bson:"origin_urls_lst"         json:"origin_urls_lst"`
	Valid_for_crawl_lst     []bool    `bson:"valid_for_crawl_lst"     json:"valid_for_crawl_lst"`  //if the link is to be crawled/followed, or should be ignored
	Fetched_lst             []bool    `bson:"fetched_lst"             json:"fetched_lst"`          //if the link's HTML was downloaded
	Images_processed_lst    []bool    `bson:"images_processed_lst"    json:"images_processed_lst"` //if the images of this links HTML page were downloaded/processed
}

type Gf_stat__unresolved_links struct {
	Origin_domain_str             string     `bson:"_id"                           json:"origin_domain_str"`
	Origin_urls_lst               []string   `bson:"origin_urls_lst"               json:"origin_urls_lst"`
	Counts__from_origin_urls_lst  []int      `bson:"counts__from_origin_urls_lst"  json:"counts__from_origin_urls_lst"`
	A_hrefs__from_origin_urls_lst [][]string `bson:"a_hrefs__from_origin_urls_lst" json:"a_hrefs__from_origin_urls_lst"`
}

type Gf_stat__links_in_day struct {
	Total_count_int           int `bson:"total_count_int"           json:"total_count_int"`
	Valid_for_crawl_total_int int `bson:"valid_for_crawl_total_int" json:"valid_for_crawl_total_int"`
	Fetched_total_int         int `bson:"fetched_total_int"         json:"fetched_total_int"`
}

//-------------------------------------------------
func stats__new_links_by_day(p_runtime_sys *gf_core.RuntimeSys) (map[string]interface{}, *gf_core.GFerror) {
	p_runtime_sys.LogFun("FUN_ENTER", "gf_crawl_stats__links.stats__new_links_by_day()")

	type Minimal_link struct {
		Creation_unix_time_f float64 `bson:"creation_unix_time_f"`
		Valid_for_crawl_bool bool    `bson:"valid_for_crawl_bool"`
		Fetched_bool         bool    `bson:"fetched_bool"`
	}

	ctx := context.Background()

	pipeline := mongo.Pipeline{
		{
			{"$match", bson.M{"t": "crawler_page_outgoing_link"}},
		},
		{
			{"$project", bson.D{
				{"creation_unix_time_f", true},
				{"valid_for_crawl_bool", true},
				{"fetched_bool",         true},
			}},
		},
		{
			{"$sort", bson.M{"creation_unix_time_f": -1}},
		},
	}

	/*pipe := p_runtime_sys.Mongo_db.Collection("gf_crawl").Pipe([]bson.M{
		bson.M{"$match": bson.M{
				"t": "crawler_page_outgoing_link",
			},
		},
		bson.M{"$project": bson.M{
				"creation_unix_time_f": true,
				"valid_for_crawl_bool": true,
				"fetched_bool":         true,
				// "id_str"               :true,
				// "cycle_run_id_str"     :true,
				// "domain_str"           :true,
				// "a_href_str"           :true, //actual link from the html <a> page ('href' parameter)
				// "origin_url_str"       :true, //page url from whos html this element was extracted
				// "images_processed_bool":true,
			},
		},
		bson.M{"$sort": bson.M{
				"creation_unix_time_f": -1,
			},
		},
	})*/

	cursor, err := p_runtime_sys.Mongo_db.Collection("gf_crawl").Aggregate(ctx, pipeline)
	if err != nil {

		gf_err := gf_core.MongoHandleError("failed to run an aggregation pipeline to get new links by day",
			"mongodb_aggregation_error",
			map[string]interface{}{},
			err, "gf_crawl_stats", p_runtime_sys)
		return nil, gf_err
	}
	defer cursor.Close(ctx)

	/*results_lst := []Minimal_link{}
	err         := pipe.AllowDiskUse().All(&results_lst)

	if err != nil {
		gf_err := gf_core.MongoHandleError("failed to run an aggregation pipeline to get new links by day",
			"mongodb_aggregation_error",
			nil, err, "gf_crawl_stats", p_runtime_sys)
		return nil, gf_err
	}*/

	results_lst := []Minimal_link{}
	for cursor.Next(ctx) {

		var r Minimal_link
		err := cursor.Decode(&r)
		if err != nil {
			gf_err := gf_core.MongoHandleError("failed to run an aggregation pipeline to get new links by day",
				"mongodb_cursor_decode",
				map[string]interface{}{},
				err, "gf_crawl_stats", p_runtime_sys)
			return nil, gf_err
		}
	
		results_lst = append(results_lst, r)
	}




	//--------------------
	// AGGREGATE DAY COUNTS - app-layer DB join
	new_links_counts_map := map[int]*Gf_stat__links_in_day{}
	keys_lst             := []int{}
	for _, l := range results_lst {

		tm                 := time.Unix(int64(l.Creation_unix_time_f), 0)
		year_day_id_int, _ := strconv.Atoi(fmt.Sprintf("%d%d", tm.Year(), tm.YearDay()))

		//--------------
		var stat_r *Gf_stat__links_in_day
		if stat, ok := new_links_counts_map[year_day_id_int]; ok {
			stat_r = stat
		} else {
			//-----------------
			// CREATE_NEW
			stat                                 := &Gf_stat__links_in_day{}
			new_links_counts_map[year_day_id_int] = stat
			stat_r                                = stat

			keys_lst = append(keys_lst, year_day_id_int)

			//-----------------
		}

		//--------------

		stat_r.Total_count_int = stat_r.Total_count_int+1

		if l.Valid_for_crawl_bool {
			stat_r.Valid_for_crawl_total_int = stat_r.Valid_for_crawl_total_int + 1
		}

		if l.Fetched_bool {
			stat_r.Fetched_total_int = stat_r.Fetched_total_int + 1
		}
	}

	//--------------------

	sort.Ints(keys_lst)

	new_links_counts__sorted_lst := []*Gf_stat__links_in_day{}
	for _, k := range keys_lst {

		stat                        := new_links_counts_map[k]
		new_links_counts__sorted_lst = append(new_links_counts__sorted_lst, stat)
	}

	fmt.Println("DONE SORTING >>>>>>>>>>>>>>>>>>>>>>>>>>>>>")
	fmt.Println(len(new_links_counts__sorted_lst))
	for _, a := range new_links_counts__sorted_lst {
		fmt.Println(a)
	}
	
	data_map := map[string]interface{}{
		"new_links_per_day_lst": new_links_counts__sorted_lst,
	}
	return data_map,nil
}

//-------------------------------------------------
func stats__unresolved_links(p_runtime_sys *gf_core.RuntimeSys) (map[string]interface{}, *gf_core.GFerror) {
	p_runtime_sys.LogFun("FUN_ENTER", "gf_crawl_stats__links.stats__unresolved_links()")



	ctx := context.Background()

	pipeline := mongo.Pipeline{
		{
			{"$match", bson.D{
				{"t",                    "crawler_page_outgoing_link"},
				{"valid_for_crawl_bool", true},
				{"fetched_bool",         false},
			}},
		},
		{
			{"$sort", bson.D{{"creation_unix_time_f", -1}}},
		},
		{
			{"$group", bson.D{
				{"_id",         bson.M{"origin_domain_str": "$origin_domain_str", "origin_url_str": "$origin_url_str",}},
				{"count_int",   bson.M{"$sum":  1}},
				{"a_hrefs_lst", bson.M{"$push": "$a_href_str"}},
			}},
		},
		{
			{"$group", bson.M{
				"_id":                           "$_id.origin_domain_str",
				"origin_urls_lst":               bson.M{"$push": "$_id.origin_url_str"},
				"counts__from_origin_urls_lst":  bson.M{"$push": "$count_int"},
				"a_hrefs__from_origin_urls_lst": bson.M{"$push": "$a_hrefs_lst"},
			}},
		},

	}

	/*pipe := p_runtime_sys.Mongo_db.Collection("gf_crawl").Pipe([]bson.M{
		bson.M{"$match": bson.M{
				"t":                    "crawler_page_outgoing_link",
				"valid_for_crawl_bool": true,
				"fetched_bool":         false,
			},
		},

		bson.M{"$sort": bson.M{
				"creation_unix_time_f": -1,
			},
		},

		bson.M{"$group": bson.M{
				"_id":         bson.M{"origin_domain_str": "$origin_domain_str", "origin_url_str": "$origin_url_str",},
				"count_int":   bson.M{"$sum":  1},
				"a_hrefs_lst": bson.M{"$push": "$a_href_str",},
			},
		},

		bson.M{"$group":bson.M{
				"_id":                           "$_id.origin_domain_str",
				"origin_urls_lst":               bson.M{"$push": "$_id.origin_url_str"},
				"counts__from_origin_urls_lst":  bson.M{"$push": "$count_int"},
				"a_hrefs__from_origin_urls_lst": bson.M{"$push": "$a_hrefs_lst"},
			},
		},
	})*/


	cursor, err := p_runtime_sys.Mongo_db.Collection("gf_crawl").Aggregate(ctx, pipeline)
	if err != nil {

		gf_err := gf_core.MongoHandleError("failed to run an aggregation pipeline to unresolved links",
			"mongodb_aggregation_error",
			map[string]interface{}{},
			err, "gf_crawl_stats", p_runtime_sys)
		return nil, gf_err
	}
	defer cursor.Close(ctx)


	/*results_lst := []Gf_stat__unresolved_links{}
	err         := pipe.AllowDiskUse().All(&results_lst)

	if err != nil {
		gf_err := gf_core.MongoHandleError("failed to run an aggregation pipeline to unresolved links",
			"mongodb_aggregation_error",
			nil, err, "gf_crawl_stats", p_runtime_sys)
		return nil, gf_err
	}*/

	results_lst := []Gf_stat__unresolved_links{}
	for cursor.Next(ctx) {

		var r Gf_stat__unresolved_links
		err := cursor.Decode(&r)
		if err != nil {
			gf_err := gf_core.MongoHandleError("failed to run an aggregation pipeline to unresolved links",
				"mongodb_cursor_decode",
				map[string]interface{}{},
				err, "gf_crawl_stats", p_runtime_sys)
			return nil, gf_err
		}
	
		results_lst = append(results_lst, r)
	}

	data_map := map[string]interface{}{
		"unresolved_links_lst": results_lst,
	}
	return data_map, nil
}

//-------------------------------------------------
func stats__crawled_links_domains(p_runtime_sys *gf_core.RuntimeSys) (map[string]interface{}, *gf_core.GFerror) {
	p_runtime_sys.LogFun("FUN_ENTER", "gf_crawl_stats__links.stats__crawled_links_domains()")



	ctx := context.Background()

	pipeline := mongo.Pipeline{
		{
			{"$match", bson.M{"t": "crawler_page_outgoing_link"}},
		},
		{
			{"$project", bson.M{
				"id_str":                true,
				"creation_unix_time_f":  true,
				"cycle_run_id_str":      true,
				"domain_str":            true,
				"a_href_str":            true, // actual link from the html <a> page ('href' parameter)
				"origin_url_str":        true, // page url from whos html this element was extracted
				"valid_for_crawl_bool":  true,
				"fetched_bool":          true,
				"images_processed_bool": true,
			}},
		},
		{
			{"$group", bson.M{
				"_id":                     "$domain_str",
				"links_count_int":         bson.M{"$sum":      1},
				"creation_unix_times_lst": bson.M{"$push":     "$creation_unix_time_f"},
				"a_href_lst":              bson.M{"$push":     "$a_href_str"},
				"origin_urls_lst":         bson.M{"$addToSet": "$origin_url_str"},
				"valid_for_crawl_lst":     bson.M{"$push":     "$valid_for_crawl_bool"},  // if the link is to be crawled/followed, or should be ignored
				"fetched_lst":             bson.M{"$push":     "$fetched_bool"},          // if the link's HTML was downloaded
				"images_processed_lst":    bson.M{"$push":     "$images_processed_bool"}, // if the images of this links HTML page were downloaded/processed
			},},
		},
		{
			{"$sort", bson.M{"links_count_int": -1}},
		},
	}



	/*pipe := p_runtime_sys.Mongo_db.Collection("gf_crawl").Pipe([]bson.M{
		bson.M{"$match": bson.M{
				"t": "crawler_page_outgoing_link",
			},
		},

		bson.M{"$project": bson.M{
				"id_str":                true,
				"creation_unix_time_f":  true,
				"cycle_run_id_str":      true,
				"domain_str":            true,
				"a_href_str":            true, //actual link from the html <a> page ('href' parameter)
				"origin_url_str":        true, //page url from whos html this element was extracted
				"valid_for_crawl_bool":  true,
				"fetched_bool":          true,
				"images_processed_bool": true,
			},
		},

		bson.M{"$group": bson.M{
				"_id":                     "$domain_str",
				"links_count_int":         bson.M{"$sum":      1},
				"creation_unix_times_lst": bson.M{"$push":     "$creation_unix_time_f"},
				"a_href_lst":              bson.M{"$push":     "$a_href_str"},
				"origin_urls_lst":         bson.M{"$addToSet": "$origin_url_str"},
				"valid_for_crawl_lst":     bson.M{"$push":     "$valid_for_crawl_bool"},  //if the link is to be crawled/followed, or should be ignored
				"fetched_lst":             bson.M{"$push":     "$fetched_bool"},          //if the link's HTML was downloaded
				"images_processed_lst":    bson.M{"$push":     "$images_processed_bool"}, //if the images of this links HTML page were downloaded/processed
			},
		},

		bson.M{"$sort": bson.M{
				"links_count_int": -1,
			},
		},
	})*/


	cursor, err := p_runtime_sys.Mongo_db.Collection("gf_crawl").Aggregate(ctx, pipeline)
	if err != nil {

		gf_err := gf_core.MongoHandleError("failed to run an aggregation pipeline to get crawled links domains",
			"mongodb_aggregation_error",
			map[string]interface{}{},
			err, "gf_images_stats", p_runtime_sys)
		return nil, gf_err
	}
	defer cursor.Close(ctx)

	/*results_lst := []Gf_stat__crawled_links_domain{}
	err         := pipe.AllowDiskUse().All(&results_lst)

	if err != nil {
		gf_err := gf_core.MongoHandleError("failed to run an aggregation pipeline to get crawled links domains",
			"mongodb_aggregation_error",
			nil, err, "gf_crawl_stats", p_runtime_sys)
		return nil, gf_err
	}*/

	results_lst := []Gf_stat__crawled_links_domain{}
	for cursor.Next(ctx) {

		var r Gf_stat__crawled_links_domain
		err := cursor.Decode(&r)
		if err != nil {
			gf_err := gf_core.MongoHandleError("failed to run an aggregation pipeline to get crawled links domains",
				"mongodb_cursor_decode",
				map[string]interface{}{},
				err, "gf_images_stats", p_runtime_sys)
			return nil, gf_err
		}
	
		results_lst = append(results_lst, r)
	}


	data_map := map[string]interface{}{
		"crawled_links_domains_lst": results_lst,
	}
	return data_map, nil
}